src/types/params.ts

```diff
@@ -102,15 +99,19 @@ export type ExtractParamTypes<TParams extends Record<string, Param>> = Identity<
  * @template TParam - The parameter type.
  * @returns The extracted type, or 'string' as a fallback.
  */
-export type ExtractParamType<TParam extends Param> = TParam extends ParamGetSet<infer Type>
-  ? TParam extends IsOptionalParam
+export type ExtractParamType<TParam extends Param, TParamKey extends PropertyKey = string> = TParam extends ParamGetSet<infer Type>
+  ? TParamKey extends `?${string}`
     ? TParam extends ParamWithDefault
       ? Type
       : Type | undefined
     : Type
   : TParam extends ParamGetter
-    ? ReturnType<TParam>
-    : string
+    ? TParamKey extends `?${string}`
+      ? ReturnType<TParam> | undefined
+      : ReturnType<TParam>
+    : TParamKey extends `?${string}`
+      ? string | undefined
+      : string
```

Comment:
previously, optional params could only be `ParamGetSet`, after this PR any param could be optional